File: 2025-12/2025-12-07/string_compression.py
def compress_string(sentence):
    words = sentence.split()
    compressed = []
    counter = 0
    i = 0

    while (i < len(words)):
        j = i
    
        while (j < len(words) and words[i] == words[j]):
            counter += 1
            j += 1

        compressed.append(f"{words[i]}({counter})" if counter > 1 else words[i])
        counter = 0
        i = j

    return " ".join(compressed)

print(compress_string("yes yes yes please"));
print(compress_string("I have have have apples"));
print(compress_string("one one three and to the the the the"));
print(compress_string("route route route route route route tee tee tee tee tee tee"));

# --- TEST SUITE ---

tests_text = '''
1. compress_string("yes yes yes please") should return "yes(3) please".
2. compress_string("I have have have apples") should return "I have(3) apples".
3. compress_string("one one three and to the the the the") should return "one(2) three and to the(4)".
4. compress_string("route route route route route route tee tee tee tee tee tee") should return "route(6) tee(6)".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+)\.$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 12 December/12-07/stringCompression.js
function compressString(sentence) {
  const words = sentence.split(" ");
  const compressed = [];
  let counter = 0;
  let i = 0;

  while (i < words.length) {
    let j = i;
	
    while (words[i] === words[j]) {
      ++counter;
      ++j;
    }
	
    compressed.push(counter > 1 ? `${words[i]}(${counter})` : words[i]);
    counter = 0;
    i = j;
  }

  return compressed.join(" ");
}

console.log(compressString("yes yes yes please"));
console.log(compressString("I have have have apples"));
console.log(compressString("one one three and to the the the the"));
console.log(compressString("route route route route route route tee tee tee tee tee tee"));

// --- TEST SUITE ---

const testsText = `
1. compressString("yes yes yes please") should return "yes(3) please".
2. compressString("I have have have apples") should return "I have(3) apples".
3. compressString("one one three and to the the the the") should return "one(2) three and to the(4)".
4. compressString("route route route route route route tee tee tee tee tee tee") should return "route(6) tee(6)".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+)\.$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);